addons/web/static/src/webclient/commands/command_service.js

```diff
@@ -9,43 +9,87 @@ import { CommandPaletteDialog } from "./command_palette_dialog";
 
 /**
  * @typedef {{
- *  name: string,
- *  action: ()=>(void | NextProvider),
- *  category?: string,
- *  hotkey?: string,
- *  hotkeyOptions?: any,
+ *  name: string;
+ *  action: ()=>(void | CommandPaletteConfig);
+ *  category?: string;
  * }} Command
  */
 
 /**
  * @typedef {{
  *  category?: string;
+ *  global?: boolean;
  *  hotkey?: string;
- *  hotkeyOptions?: any;
- * }} CommandServiceAddOptions
+ * }} CommandOptions
+ */
+
+/**
+ * @typedef {Command & CommandOptions & {
+ *  activeElement?: HTMLElement;
+ *  removeHotkey?: ()=>void;
+ * }} CommandRegistration
  */
 
+const commandCategoryRegistry = registry.category("command_categories");
+const commandEmptyMessageRegistry = registry.category("command_empty_list");
+const commandProviderRegistry = registry.category("command_provider");
+
 export const commandService = {
     dependencies: ["dialog", "hotkey", "ui"],
     start(env, { dialog, hotkey: hotkeyService, ui }) {
+        /** @type {Map<CommandRegistration>} */
         const registeredCommands = new Map();
         let nextToken = 0;
         let isPaletteOpened = false;
 
-        hotkeyService.add("control+k", openPalette, {
+        hotkeyService.add("control+k", openMainPalette, {
             global: true,
         });
 
-        function openPalette() {
+        function openMainPalette(namespace) {
+            const categoriesByNamespace = {};
+            commandCategoryRegistry.getEntries().forEach(([category, el]) => {
+                const namespace = el.namespace ? el.namespace : "default";
+                if (namespace in categoriesByNamespace) {
+                    categoriesByNamespace[namespace].push(category);
+                } else {
+                    categoriesByNamespace[namespace] = [category];
+                }
+            });
+
+            const emptyMessageByNamespace = {};
+            commandEmptyMessageRegistry.getEntries().forEach(([key, message]) => {
+                emptyMessageByNamespace[key] = message.toString();
+            });
+
+            const config = {
+                categoriesByNamespace,
+                emptyMessageByNamespace,
+                footer: env
+                    ._t(
+                        "<span class='o_promote'>TIP</span> — search for <span class='o_promote'>@</span>users, <span class='o_promote'>#</span>channels, and <span class='o_promote'>/</span>menus"
+                    )
+                    .toString(),
+                footerIsHtml: true,
```

Comment:
Why the `toString`? Also is it actually necessary? This seems to be the lone use of the `footerIsHtml` feature, couldn't it be solved using a template object / name or something? IIRC owl supports inline-ish templates, maybe that'd be an option? Instead of text content the template would be passed a template name or object or whatever owl does / supports, and would call that.

Reply:
Good idea. Thx